### ISSUE TITLE
Scalar types return wrong error when conversion from string fails (migrated from Trac #99)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#706]
Original ticket http://projects.scipy.org/numpy/ticket/99
Reported 2006-05-07 by atmention:alberts, assigned to atmention:teoliphant.

Works for `int`, breaks for `intp` when it shouldn't, and with the wrong error:
    In [32]: int('0xb72a7008', 16)
    Out[32]: 3073011720L

```
In [33]: N.intp('0xb72a7008', 16)
TypeError: function takes exactly 1 argument (2 given)
```

`0xb72a7008` was actually a string returned in by a !NumPy array's `__array_data__[0]` on Linux.

Breaks for `long` and `intp`, but `intp` raises the wrong error:
    In [34]: int('0x1', 32)
    ValueError: invalid literal for int(): 0x1

```
In [35]: N.intp('0x1', 32)
TypeError: function takes exactly 1 argument (2 given)
```

Tested with r2476.

Please add a unit test when fixing this bug. Thanks!

**Comment by thouis** [comment]
Comment in Trac by atmention:alberts, 2006-05-07

Workaround for dealing with addresses:

```
a = N.array([1.])
N.cast[N.intp](int(a.__array_data__[0], 16))
```

However, the bugs remain valid. It also isn't clear what intp should do. I'd expect to be able to convert an address to intp without casting.

**Comment by thouis** [comment]
Comment in Trac by atmention:teoliphant, 2006-05-09

You are assuming too much for the N.intp data-type.  As the error shows, it does not take two arguments.  

It is not unreasonable that the integer constructors would behave similarly to Python but this is not the current behavior.   The constructors are meant to produce arrays of the given data-type.

**Comment by thouis** [comment]
Comment in Trac by atmention:alberts, 2006-05-09

If `intp` doesn't take two arguments, how does this work?

```
In [6]: N.intp('0xf', 16)
Out[6]: 15
```

**Comment by thouis** [comment]
Comment in Trac by atmention:teoliphant, 2006-07-20

Array scalars that inherit from a Python type (string, unicode, float, complex) and possibly (intp, int_, long_, longlong_) give the underlying Python code a chance to figure out the input first.  If this fails, then the default code gets executed (where the number of arguments error is occurring).  

I've changed this code to not move on if the number of arguments is larger than 1 (what is expected in the next code segment).  

Changed in r2854
